catalog/cataloger_export.go

```diff
@@ -0,0 +1,139 @@
+	rows, err := c.db.Query(
+		`SELECT r.name repository, b.name branch,
+                     e.export_path export_path, e.export_status_path export_status_path,
+                     e.last_keys_in_prefix_regexp last_keys_in_prefix_regexp
+                 FROM catalog_branches_export e JOIN catalog_branches b ON e.branch_id = b.branch_id
+                      JOIN catalog_repositories r ON b.repository_id = r.id`)
+	if err != nil {
+		return nil, err
+	}
+	for rows.Next() {
+		var rec ExportConfigurationForBranch
+		if err = rows.Scan(&rec); err != nil {
+			return nil, fmt.Errorf("scan configuration %+v: %w", rows, err)
+		}
+		ret = append(ret, rec)
+	}
+	return ret, nil
+}
+
+func (c *cataloger) PutExportConfiguration(repository string, branch string, conf *ExportConfiguration) error {
+	_, err := c.db.Transact(func(tx db.Tx) (interface{}, error) {
+		branchID, err := c.getBranchIDCache(tx, repository, branch)
+		if err != nil {
+			return nil, err
+		}
+		_, err = c.db.Exec(
+			`INSERT INTO catalog_branches_export (
+                             branch_id, export_path, export_status_path, last_keys_in_prefix_regexp)
+                         VALUES ($1, $2, $3, $4)
+                         ON CONFLICT (branch_id)
+                         DO UPDATE SET (branch_id, export_path, export_status_path, last_keys_in_prefix_regexp) =
+                             (EXCLUDED.branch_id, EXCLUDED.export_path, EXCLUDED.export_status_path, EXCLUDED.last_keys_in_prefix_regexp)`,
+			branchID, conf.Path, conf.StatusPath, conf.LastKeysInPrefixRegexp)
+		return nil, err
+	})
+	return err
+}
+
+const (
+	groupStart = "(?:"
+	groupEnd   = ")"
+)
+
+// DisjunctRegexps returns a single regexp holding the disjunction ("|") of regexps.
+func DisjunctRegexps(regexps []string) (*regexp.Regexp, error) {
+	parts := make([]string, len(regexps))
+	for i, re := range regexps {
+		_, err := syntax.Parse(re, syntax.Perl)
+		if err != nil {
+			return nil, fmt.Errorf("%s: %w", re, err)
+		}
+		parts[i] = groupStart + re + groupEnd
+	}
+	d := strings.Join(parts, "|")
+	return regexp.Compile(d)
+}
+
+var (
+	ErrNotBracketed = errors.New("not a bracketed string")
+	unbracketRegexp = regexp.MustCompile("^" + regexp.QuoteMeta(groupStart) + "(.*)" + regexp.QuoteMeta(groupEnd) + "$")
+)
+
+func unbracket(s string) (string, error) {
+	sub := unbracketRegexp.FindStringSubmatch(s)
+	if len(sub) == 0 {
+		return "", fmt.Errorf("%s: %w", s, ErrNotBracketed)
+	}
+	return sub[1], nil
+}
+
+// DeconstructDisjunction returns the text forms of the regexps in the disjunction rex.  rex
+// should be constructed (only) by DisjunctRegexps.
+func DeconstructDisjunction(regexp *regexp.Regexp) ([]string, error) { // nolint:interfacer
+	// Why nollnt above? I really do want this regexp handling function to take a regexp, not
+	// expvar.Var hich is a silly alias to Stringer.
+
+	s := regexp.String()
+	if len(s) == 0 {
+		return nil, nil
+	}
+	regexpParts := strings.Split(s, "|")
```

Comment:
Known issue (working on it while you review): This split is much too deep, e.g if you disjunct "a|b" and "c" it breaks.

Reply:
Fixed this, got rid of all the de-disjunctioning.